libexec/python/docker/api.py

```diff
@@ -163,19 +163,27 @@ def update_token(self, response=None, auth=None):
                 sys.exit(1)
 
             challenge = response.headers["Www-Authenticate"]
-            regexp = '^Bearer\s+realm="(.+)",service="(.+)",scope="(.+)",?'
+            regexp = '^Bearer\s+realm="(.+?)"'
+            regexp += '(?:,service="(.+?)")?'
+            regexp += '(?:,scope="(.+?)")?'
             match = re.match(regexp, challenge)
 
             if not match:
                 bot.error("Unrecognized authentication challenge, exiting.")
                 sys.exit(1)
 
             realm = match.group(1)
-            service = match.group(2)
-            scope = match.group(3).split(',')[0]
+            service = match.group(2) or ''
+            scope = (match.group(3) or '').split(',')[0]
 
-            self.token_url = ("%s?service=%s&expires_in=9000&scope=%s"
-                              % (realm, service, scope))
+            token_url_params = ['expires_in=%s' % expires_in]
+
+            if service:
+                token_url_params.append('service=%s' % service)
+            if scope:
+                token_url_params.append('scope=%s' % scope)
+
+            self.token_url = "%s?%s" % (realm, '&'.join(token_url_params))
```

Comment:
Here we are going to end up with realm,expires_in,service,scope which wouldn't match the regex mentioned in previous comment due to ordering - which is what first made me think about ordering above.

If this works with repositories out there, then they don't expect strict realm,service,scope - so maybe this code should not? I could be wrong - but if order is important is there any reference out there? I had a quick look but couldn't find anything.

Reply:
@dctrud It's somewhere in this tangle --> https://tools.ietf.org/html/rfc6750#section-3

> All challenges defined by this specification MUST use the auth-scheme value "Bearer".  This scheme MUST be followed by one or more auth-param values.  The auth-param attributes used or defined by this specification are as follows.  Other auth-param attributes MAY be used as well.

So (overall) I (think?) the other values come after Bearer, but it would also depend on if other repositories are implementing a different kind of auth too. It's quite a messy space! Anyhoo, hopefully some digging in there will give a satisfactory answer!

Reply:
@vsoch - yup I agree the come after bearer - it's just i'm not clear that the ones after bearer are required to be in fixed order. The regex is enforcing the order is `realm,service,scope` what if we saw `realm,scope,service`.

Reply:
The order of query params in the token URL really shouldn't matter - query params get parsed as a hash by every lib I've seen although I could be wrong...